Libplanet/Crypto/SymmetricKey.cs

```diff
@@ -0,0 +1,209 @@
+        /// It returns <c>null</c> if the <paramref name="ciphertext"/> is
+        /// invalid (this behavior will be eventually changed in the future to
+        /// throw an exception instead).</returns>
+        /// <remarks>
+        /// Although the parameter name <paramref name="ciphertext"/> has the
+        /// word &#x201c;text&#x201d;, both a <paramref name="ciphertext"/>
+        /// and a returned message are a <see cref="byte"/> string,
+        /// not a Unicode <see cref="string"/>.
+        /// </remarks>
+        /// <seealso cref="Encrypt(byte[], byte[])"/>
+        [Pure]
+        public byte[] Decrypt(byte[] ciphertext, int nonSecretLength = 0)
+        {
+            if (ciphertext == null || ciphertext.Length == 0)
+            {
+                throw new ArgumentException(
+                    "Encrypted Message Required!",
+                    nameof(ciphertext));
+            }
+
+            using (var cipherStream = new MemoryStream(ciphertext))
+            using (var cipherReader = new BinaryReader(cipherStream))
+            {
+                byte[] nonSecretPayload = cipherReader.ReadBytes(
+                    nonSecretLength
+                );
+                byte[] nonce = cipherReader.ReadBytes(NonceBitSize / 8);
+
+                var cipher = new GcmBlockCipher(new AesEngine());
+                var parameters = new AeadParameters(
+                    new KeyParameter(_key),
+                    MacBitSize,
+                    nonce,
+                    nonSecretPayload);
+                cipher.Init(false, parameters);
+
+                byte[] cipherText = cipherReader.ReadBytes(
+                    ciphertext.Length - nonSecretLength - nonce.Length
+                );
+                var plainText =
+                    new byte[cipher.GetOutputSize(cipherText.Length)];
+
+                try
+                {
+                    int len = cipher.ProcessBytes(
+                        cipherText, 0, cipherText.Length, plainText, 0
+                    );
+                    cipher.DoFinal(plainText, len);
+                    return plainText;
+                }
+                catch (InvalidCipherTextException)
+                {
+                    // FIXME
+                    return null;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Gets a mutable <see cref="byte"/> array which encodes this key.
+        /// </summary>
+        /// <returns>A new mutable <see cref="byte"/> array which encodes this
+        /// key.  Since it is created every time the method is called, any
+        /// mutation on that does not affect this <see cref="SymmetricKey"/>
+        /// object.</returns>
+        /// <seealso cref="ByteArray"/>
+        /// <seealso cref="SymmetricKey(byte[])"/>
+        [Pure]
+        public byte[] ToByteArray()
+        {
+            byte[] copied = new byte[_key.Length];
```

Comment:
Can we re-use `ByteArray` here? (or, is it performance related decision?)

Reply:
It was a concern about performance thing, but I will rewrite it to use `.ToArray()` which is more readable.